rfc/0084-app-service-level-telemetry.md

```diff
@@ -0,0 +1,517 @@
+
+  This endpoint must not have authentication, and should support WebSocket
+  protocol.
+
+## Network Interacation
+
+Once the SDK determined the endpoint it should establish HTTP connection and
+upgrade it to WebSocket.
+
+The server allows to close connection at any point of time.
+
+Once the server closes connection, the SDK should pick next available endpoint.
+Both initial and next endpoints must be randomized in case of more than one
+endpoint configured.
+
+The SDK must wait for backoff interval before connection attempt to avoid
+stressing the server. The SDK might use `app_telemetry_backoff` option to allow
+to change the backoff value.
+
+The SDK must either implement or enable WebSocket PING frame, and respond to
+server request with WebSocket PONG whenever the server requests it.
+
+## Wire Protocol
+
+The structure of the protocol is intentionally simple, and yet allows for
+future extension.
+
+The SDK leverages WebSocket framing, so does not need to have special fields
+for the payload size.
+
+First byte is reserved for Command codes in requests from the server, and for
+Status for SDK responses.
+
+This RFC defines the following types of the commands.
+
+| Name            | Value  | Description                                                                  |
+|-----------------|--------|------------------------------------------------------------------------------|
+| `GET_TELEMETRY` | `0x00` | Retrieve Application Telemetry from SDK in Prometheus Text Exposition Format |
+
+And the following statuses:
+
+| Name              | Value  | Description                                                |
+|-------------------|--------|------------------------------------------------------------|
+| `SUCCESS`         | `0x00` | No errors has been occured                                 |
+| `UNKNOWN_COMMAND` | `0x01` | The SDK does not support operation requested by the server |
+
+
+### `GET_TELEMETRY`
+
+The request must not have payload.
+
+The response might have payload if the SDK has metrics to report.
+
+The SDK must reset local counters after successful scheduling response for the
+network delivery.
+
+The format of the response follows the [Text-based exposition format of the
+Prometheus](https://prometheus.io/docs/instrumenting/exposition_formats/#text-based-format),
+and includes metrics and histograms collected by the SDK since the previous
+report.
+
+#### Metrics
+
+    sdk_kv_r_utimedout{agent="couchbase-net-sdk/2.4.5.0 (clr/4.0.30319.42000) (os/Microsoft Windows NT 10.0.16299.0)", id="66388CF5BFCF7522/18CC8791579B567C", node="node1", bucket="travel-sample"} 1 1695747260
+
+Where
+
+* `sdk_kv_r_utimedout` is a name of the metric, one of the:
+
+  | Metric | Description |
+  | ---- | ---- |
+  | `sdk_{service}_r_utimedout` | Unambiguous timeout |
+  | `sdk_{service}_r_atimedout` | Ambiguous timeout |
+  | `sdk_{service}_r_canceled` | Canceled |
+  | `sdk_{service}_r_total` | Total number of operations |
```

Comment:
May the SDK omit a metric if the metric value is `0`?

Reply:
I think at this time it would be better to keep the zero metric, because it also carries timestamp, so the collector will be able to plot zero at this time point.

Reply:
Updated the RFC